### PR TITLE
日々の進捗に応じてgraphに色を塗る

### PR DESCRIPTION
**Pull request by highwide**
- [x] nヶ月分の進捗を取得するメソッドをModelに追加
- [x] 日々の進捗を整形するメソッドをきちんとViewが求める感じに修正
- [x] 進捗に応じて色塗りをするメソッドをHelperに追加
- [ ] ~~テスト書かねば...(´・ω・｀)~~

testは別branchで対応

### CODE REVIEW

File: app/helpers/practice_helper.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+module PracticeHelper
+  def color(daily)
```
Comment (highwide) [inline app/helpers/practice_helper.rb]: __[rails_best_practices]__
* remove unused methods (PracticeHelper#color)

File: app/controllers/practices_controller.rb [hunk 2]
```diff
@@ -34,19 +36,20 @@ def update
...
+def format(progresses, months)
   end_date = Time.zone.now
-  start_date = end_date.months_ago(month)
+  start_date = end_date.months_ago(months)
   date_per_week = []
 
   while (start_date <= end_date)
-    if date_per_week.empty? || date_per_week.last.last.wday == 6
-      date_per_week << [start_date]
+    if date_per_week.empty? || date_per_week.last.any? {|k, _| k.wday == 6}
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: __[reek]__
* [DuplicateMethodCall] calls date_per_week.last 2 times
* [UncommunicativeVariableName] has the variable name 'k'

File: app/controllers/practices_controller.rb [hunk 3]
```diff
@@ -34,19 +36,20 @@ def update
...
+      date_per_week << \
+          {start_date => progresses.select{|p| p.done_at == start_date.to_date}.last}
     else
-      date_per_week.last << start_date
+      date_per_week.last[start_date] = progresses.select{|p| p.done_at == start_date.to_date}.last
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: __[reek]__
* [DuplicateMethodCall] calls date_per_week.last 2 times
* [DuplicateMethodCall] calls p.done_at 2 times
* [DuplicateMethodCall] calls p.done_at == start_date.to_date 2 times
* [DuplicateMethodCall] calls progresses.select 2 times
* [DuplicateMethodCall] calls progresses.select do |p| ... end 2 times
* [DuplicateMethodCall] calls progresses.select do |p| ... end.last 2 times
* [DuplicateMethodCall] calls start_date.to_date 2 times
* [UncommunicativeVariableName] has the variable name 'p'

File: app/controllers/practices_controller.rb [hunk 4]
```diff
@@ -34,19 +36,20 @@ def update
...
-  start_date = end_date.months_ago(month)
+  start_date = end_date.months_ago(months)
   date_per_week = []
 
   while (start_date <= end_date)
-    if date_per_week.empty? || date_per_week.last.last.wday == 6
-      date_per_week << [start_date]
+    if date_per_week.empty? || date_per_week.last.any? {|k, _| k.wday == 6}
+      date_per_week << \
+          {start_date => progresses.select{|p| p.done_at == start_date.to_date}.last}
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: __[reek]__
* [DuplicateMethodCall] calls p.done_at 2 times
* [DuplicateMethodCall] calls p.done_at == start_date.to_date 2 times
* [DuplicateMethodCall] calls progresses.select 2 times
* [DuplicateMethodCall] calls progresses.select do |p| ... end 2 times
* [DuplicateMethodCall] calls progresses.select do |p| ... end.last 2 times
* [DuplicateMethodCall] calls start_date.to_date 2 times
* [UncommunicativeVariableName] has the variable name 'p'

File: app/controllers/practices_controller.rb [hunk 5]
```diff
@@ -34,19 +36,20 @@ def update
 end
 
 private
-def format_dates_by_term(month)
+
+def format(progresses, months)
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: **[reek]**
- [UtilityFunction] doesn't depend on instance state
- [TooManyStatements] has approx 10 statements

File: app/helpers/practice_helper.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    if daily.blank? || daily.progress == 0
```
Comment (highwide) [inline app/helpers/practice_helper.rb]: **[reek]**
- [DuplicateMethodCall] calls daily.progress 2 times

File: app/helpers/practice_helper.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      '#eee'
+    else
+      progress_rate = daily.progress.to_f / daily.goal.to_f
```
Comment (highwide) [inline app/helpers/practice_helper.rb]: **[reek]**
- [DuplicateMethodCall] calls daily.progress 2 times

File: app/helpers/practice_helper.rb (hunk 1 above)
Comment (highwide) [inline app/helpers/practice_helper.rb:3]: **[reek]**
- [UtilityFunction] doesn't depend on instance state
- [TooManyStatements] has approx 7 statements

File: app/models/daily_progress.rb [hunk 8]
```diff
@@ -19,4 +19,11 @@ class DailyProgress < ActiveRecord::Base
   belongs_to :practice
   validates  :practice,    presence: true
   validates  :done_at,     presence: true
+
+  def self.in_this_months(month)
```
Comment (ta1kt0me) [inline app/models/daily_progress.rb]: selfいらなそう

Comment (highwide) [inline app/models/daily_progress.rb]: 話し合いの結果、scopeの方がいいのでは、という結論に。

File: app/models/daily_progress.rb [hunk 9]
```diff
@@ -19,4 +19,11 @@ class DailyProgress < ActiveRecord::Base
...
+    end_date = Time.zone.now
+    start_date = end_date.months_ago(month)
+
+    where(done_at: start_date..end_date)
```
Comment (ta1kt0me) [inline app/models/daily_progress.rb]:
```
where("done_at >= ?",Time.zone.now.months_ago(12))
```

でどうですか？

Comment (highwide) [inline app/models/daily_progress.rb]: ++

File: app/helpers/practice_helper.rb (hunk 6 above)
Comment (ta1kt0me) [inline app/helpers/practice_helper.rb]:
```
daily.try(:progress) == 0
```

File: app/controllers/practices_controller.rb (hunk 5 above)
Comment (ta1kt0me) [inline app/controllers/practices_controller.rb]: クラスの外に出てしまっている...! :scream:

File: app/helpers/practice_helper.rb (hunk 6 above)
Comment (highwide) [inline app/helpers/practice_helper.rb]: progress が 0だったらどうせ後の progress_rate が0になる

File: app/controllers/practices_controller.rb (hunk 5 above)
Comment (highwide) [inline app/controllers/practices_controller.rb]: やばいｗすみません :innocent:

File: app/controllers/practices_controller.rb [hunk 10]
```diff
@@ -34,19 +36,20 @@ def update
...
-    if date_per_week.empty? || date_per_week.last.last.wday == 6
-      date_per_week << [start_date]
+    if date_per_week.empty? || date_per_week.last.any? {|k, _| k.wday == 6}
+      date_per_week << \
+          {start_date => progresses.select{|p| p.done_at == start_date.to_date}.last}
     else
-      date_per_week.last << start_date
+      date_per_week.last[start_date] = progresses.select{|p| p.done_at == start_date.to_date}.last
     end
     start_date += 1.day
```
Comment (ta1kt0me) [inline app/controllers/practices_controller.rb]: こうするとこの行がいらなさそう

```
  (start_date..end_date).each do |day|
    if date_per_week.empty? || date_per_week.last.any? {|k, _| k.wday == 6}
      date_per_week << \
          {day => progresses.select{|p| p.done_at == day.to_date}.last}
    else
      date_per_week.last[day] = progresses.select{|p| p.done_at == day.to_date}.last
    end
  end
```

File: app/controllers/practices_controller.rb (hunk 4 above)
Comment (ta1kt0me) [inline app/controllers/practices_controller.rb]: findだけで良いなら、こうできそう。

```
    progress = progresses.find{|p| p.done_at == day.to_date}
    if date_per_week.empty? || date_per_week.last.any? {|k, _| k.wday == 6}
      date_per_week << {day => progress}
    else
      date_per_week.last[day] = progress
    end
```

File: app/controllers/practices_controller.rb [hunk 11]
```diff
@@ -34,19 +36,20 @@ def update
...
   end
-
   date_per_week
```
Comment (ta1kt0me) [inline app/controllers/practices_controller.rb]: each_with_objectを使うと、date_per_weekを消せそう。

```
(1..2).each_with_object([]) {|i, result| result << i }
```

File: app/controllers/practices_controller.rb (hunk 10 above)
Comment (highwide) [inline app/controllers/practices_controller.rb]: 素敵 :heart_eyes:

File: app/controllers/practices_controller.rb (hunk 11 above)
Comment (ta1kt0me) [inline app/controllers/practices_controller.rb]: http://ruby-doc.org/core-2.2.2/Enumerable.html

File: app/helpers/practice_helper.rb (hunk 6 above)
Comment (ta1kt0me) [inline app/helpers/practice_helper.rb]: ネスト深くなるので、returnしちゃっても良いのでは。

File: app/helpers/practice_helper.rb [hunk 12]
```diff
@@ -0,0 +1,18 @@
+module PracticeHelper
+  # TODO: 使ってみて緑の濃さ調整する
+  def color(daily)
```
Comment (highwide) [inline app/helpers/practice_helper.rb]: __[rails_best_practices]__
* remove unused methods (PracticeHelper#color)

File: app/controllers/practices_controller.rb [hunk 13]
```diff
@@ -31,22 +32,24 @@ def edit
...
+  # Viewで表示するgraphのためにデータ構造を整形するメソッド
+  # 日付をkey, DailyProgressモデルをvalueとするHashを作り、
+  # 1週間ごと(土曜日(wdayが6)で切り替え)に別HashとしてArrayに詰め込んでいる
+  #
+  def format_for_graph_of_a_year(progresses)
+    end_date = Time.zone.now.to_date
+    start_date = end_date.years_ago(1).to_date
+    (start_date..end_date).each_with_object([]) do |date, formated_dates|
+      progress = progresses.find{|p| p.done_at == start_date}
+      if formated_dates.empty? || formated_dates.last.any? {|k, _| k.wday == 6}
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: __[reek]__
* [DuplicateMethodCall] calls formated_dates.last 2 times
* [NestedIterators] contains iterators nested 2 deep
* [UncommunicativeVariableName] has the variable name 'k'

File: app/controllers/practices_controller.rb [hunk 14]
```diff
@@ -31,22 +32,24 @@ def edit
...
+        formated_dates << {date => progress}
+      else
+        formated_dates.last[date] = progress
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: __[reek]__
* [DuplicateMethodCall] calls formated_dates.last 2 times

File: app/controllers/practices_controller.rb [hunk 15]
```diff
@@ -31,22 +32,24 @@ def edit
...
-    else
-      date_per_week.last << start_date
+  private
+
+  #
+  # Viewで表示するgraphのためにデータ構造を整形するメソッド
+  # 日付をkey, DailyProgressモデルをvalueとするHashを作り、
+  # 1週間ごと(土曜日(wdayが6)で切り替え)に別HashとしてArrayに詰め込んでいる
+  #
+  def format_for_graph_of_a_year(progresses)
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: __[reek]__
* [UtilityFunction] doesn't depend on instance state
* [TooManyStatements] has approx 8 statements

File: app/controllers/practices_controller.rb [hunk 16]
```diff
@@ -31,22 +32,24 @@ def edit
...
+    end_date = Time.zone.now.to_date
+    start_date = end_date.years_ago(1).to_date
+    (start_date..end_date).each_with_object([]) do |date, formated_dates|
+      progress = progresses.find{|p| p.done_at == start_date}
```
Comment (highwide) [inline app/controllers/practices_controller.rb]: __[reek]__
* [UncommunicativeVariableName] has the variable name 'p'

File: app/helpers/practice_helper.rb (hunk 12 above)
Comment (highwide) [inline app/helpers/practice_helper.rb]: __[reek]__
* [UtilityFunction] doesn't depend on instance state
* [TooManyStatements] has approx 7 statements

**Comment by ta1kt0me** [comment]
よいのでは！！

**Comment by ta1kt0me** [comment]
テスト用のissue立てて、クローズでいいんじゃないでしょうか。

**Comment by highwide** [comment]
とみーさんにいろいろコメントいただいたおかげで、大変よい感じになりました！
ありがとうございます！！！

**Pull request merged**